pages/datamodel.vue

```diff
@@ -0,0 +1,435 @@
+      for( const auto& surface : my_brep.surfaces() )
+      {
+      // do something with surface (which is a Surface3D)
+      auto nbv = surface.mesh().nb_vertices();
+      }
+    </code>
+    <!-- eslint-enable -->
+
+    <p :class="paragraph">
+      Notice that each surface is get using a const reference
+      since a
+      <span v-html="surface3d" /> is not copyable nor modifiable.
+    </p>
+
+    <p :class="paragraph">
+      Now, you may want to process only a single
+      <span v-html="surface3d" />.
+      Hopefully, it is possible to get one
+      <span v-html="surface3d" /> from a
+      <span v-html="brep" />.
+      A major feature of OpenGeode is that model components are not continuously numbered:
+      they are identified by an unique index
+      <span
+        v-html="uuid"
+      />.
+      This is a fondamental design element to allow modification of OpenGeode models.
+      You can thus get a
+      <span
+        v-html="surface3d"
+      /> from its
+      <span v-html="uuid" />:
+    </p>
+
+    <!-- eslint-disable -->
+    <code class="cpp primary--text">
+      BRep my_brep;
+      uuid surf_id;
+      const auto& surface = my_brep.surface( surf_id );
+    </code>
+    <!-- eslint-enable -->
+
+    <h2 :class="section">What about other Components?</h2>
+
+    <p :class="paragraph">
+      We have seen that
+      <span v-html="brep" /> is composed of
+      <span v-html="surfaces3d" />.
+      These Surfaces delimit volumes regionalizing an object.
+      As a consequence, a
+      <span
+        v-html="brep"
+      /> is also composed of
+      <span v-html="blocks3d" />.
+      The
+      <span v-html="brep" /> class inherits from
+      <span v-html="surfaces3d" /> and
+      <span v-html="blocks3d" />.
+    </p>
+
+    <!-- eslint-disable -->
+    <code class="cpp primary--text">
+      class BRep : public Surfaces3D,
+      public Blocks3D
+      {
+      ...
+      }
+    </code>
+    <!-- eslint-enable -->
+
+    <p :class="paragraph">
+      You can count number of Blocks, iterates on Blocks,
+      get a specific
+      <span
+        v-html="block3d"
+      /> from its
+      <span v-html="uuid" /> as done for Surfaces.
+    </p>
+
+    <img src="@/assets/datamodel-3.svg" />
+
+    <p :class="paragraph">
+      But as Surfaces bound Blocks, BRep Surfaces are bounded by Lines (in blue),
+      and BRep Lines are bounded by Corners (in green). These class are named
+      <strong>Components</strong>.
+      The
+      <span v-html="brep" /> class thus inherits from all these classes.
+      To ease the conception of custom OpenGeode models, the helper class
+      <span
+        v-html="addComponents"
+      /> is provided.
+      The
+      <span v-html="brep" /> class is defined as following:
+    </p>
+
+    <img src="@/assets/datamodel-4.svg" />
+
+    <!-- eslint-disable -->
+    <code class="cpp primary--text">
+      class BRep : public AddComponents< 3, Corners, Lines, Surfaces, Blocks >
```

Comment:
I would have used a composition design pattern. Or maybe you use a decorator design? I have not looked at the code yet. Maybe I miss something.

Reply:
Here is a mixin inheritance which can be considered as a decorator. In my memories, Decorator is a bit problematic since the type of the object corresponds to the type of the Decorator and not the Base object. Am I right?